### PR TITLE
Introduce slint::run_event_loop_until_quit

### LINKED ISSUE #1499: `slint::ComponentHandle::run` returns when `on_close_requested` callback completes
I was looking to crate a system tray thing for my application and i couldn't because of this issue. even `slint::ComponentHandle::hide` makes `slint::ComponentHandle::run` return.

# Example
https://gist.github.com/qw0rd/b9e86a99129b31e502bd9e45d146a84e

### PR DESCRIPTION
**Pull request by ogoffart**
Closes #1499

### CODE REVIEW

File: api/cpp/include/slint.h [hunk 1]
```diff
@@ -1261,13 +1261,29 @@ struct [[deprecated]] VersionCheckHelper
...
+/// Enum for the event loop mode parameter of the slint::run_event_loop() function.
+/// It is used to determine when the event loop quits
+enum class EventLoopMode {
+    /// The event loop will quit when the last window is closed
+    /// or when slint::quit_event_loop() is called.
+    QuitOnLastWindowClose,
+
+    /// The event loop will keep running until slint::quit_event_loop() is called,
+    /// even when all windows are closed.
+    KeepAlive
```
Comment (tronical) [inline api/cpp/include/slint.h:1273]: I suggest to name it like in Rust.

```suggestion
    RunUntilQuit
```

File: api/cpp/lib.rs [hunk 2]
```diff
@@ -49,8 +49,16 @@ pub unsafe extern "C" fn slint_ensure_backend() {
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn slint_run_event_loop() {
-    with_platform(|b| b.run_event_loop()).unwrap();
+/// Enters the main event loop.
+pub extern "C" fn slint_run_event_loop(quit_on_last_window_close: bool) {
```
Comment (tronical) [inline api/cpp/lib.rs:53]:
```suggestion
pub extern "C" fn slint_run_event_loop(quit_on_last_window_closed: bool) {
```

File: api/cpp/lib.rs [hunk 3]
```diff
@@ -49,8 +49,16 @@ pub unsafe extern "C" fn slint_ensure_backend() {
...
+    with_platform(|b| {
+        if !quit_on_last_window_close {
```
Comment (tronical) [inline api/cpp/lib.rs:55]:
```suggestion
        if !quit_on_last_window_closed {
```

File: api/cpp/tests/platform_eventloop.cpp [hunk 4]
```diff
@@ -258,3 +269,96 @@ TEST_CASE("Blocking Event from thread")
     REQUIRE(called == 42);
     t.join();
 }
+
+#if defined(SLINT_FEATURE_INTERPRETER) && defined(SLINT_FEATURE_RENDERER_SOFTWARE)
+
+#    include <slint-interpreter.h>
+
+TEST_CASE("Close on last window closed")
```
Comment (tronical) [inline api/cpp/tests/platform_eventloop.cpp:277]:
```suggestion
TEST_CASE("Quit on last window closed")
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -15,6 +15,7 @@ All notable changes to this project are documented in this file.
 - Fixed ListView not updating when model changes (#3125)
 - Fixed TextInput on Plasma/Wayland receiveng many empty events causing selection to be cleared (#4148)
 - Added API to programmatically show a window in fullscreen mode (C++/Rust: `Window::set_fullscreen(bool)`, Node.js: `window.fullscreen`)
+- Added API to keep the event loop alive when the last window is closed (#1499).
```
Comment (tronical) [inline CHANGELOG.md:18]:
```suggestion
- Added API to keep the event loop alive when the last window is closed (#1499). (Rust: `slint::run_event_loop_until_quit(); C++: argument to `slint::run_event_loop()`)
```

Comment (tronical) [reply CHANGELOG.md:18]: JavaScript is missing though, in the implementation. If you don't plan on doing it, let me know or file an issue for 1.4.

File: api/cpp/tests/platform_eventloop.cpp [hunk 6]
```diff
@@ -258,3 +269,96 @@ TEST_CASE("Blocking Event from thread")
...
+            instance2->hide();
+            instance->hide();
+            slint::Timer::single_shot(std::chrono::milliseconds(0), [&]() {
+                instance2->show();
+                slint::quit_event_loop();
+                ok = 4;
+            });
+        });
+    });
+    slint::run_event_loop(slint::EventLoopMode::KeepAlive);
```
Comment (tronical) [inline api/cpp/tests/platform_eventloop.cpp:359]:
```suggestion
    slint::run_event_loop(slint::EventLoopMode::RunUntilQuit);
```

**Review by tronical — APPROVED** [review]

File: api/cpp/include/slint.h [hunk 7]
```diff
@@ -1261,13 +1261,29 @@ struct [[deprecated]] VersionCheckHelper
...
+
 /// Enters the main event loop. This is necessary in order to receive
 /// events from the windowing system in order to render to the screen
 /// and react to user input.
-inline void run_event_loop()
+///
+/// The mode parameter determines the behavior of the event loop when all windows are closed.
+/// By default, it is set to QuitOnLastWindowClose, which means the event loop will
+/// quit when the last window is closed.
+inline void run_event_loop(EventLoopMode mode = EventLoopMode::QuitOnLastWindowClose)
```
Comment (ogoffart) [inline api/cpp/include/slint.h:1283]:
```suggestion
inline void run_event_loop(EventLoopMode mode = EventLoopMode::QuitOnLastWindowClosed)
```

File: api/cpp/include/slint.h [hunk 8]
```diff
@@ -1261,13 +1261,29 @@ struct [[deprecated]] VersionCheckHelper
...
 {
     private_api::assert_main_thread();
-    cbindgen_private::slint_run_event_loop();
+    cbindgen_private::slint_run_event_loop(mode == EventLoopMode::QuitOnLastWindowClose);
```
Comment (ogoffart) [inline api/cpp/include/slint.h:1286]:
```suggestion
    cbindgen_private::slint_run_event_loop(mode == EventLoopMode::QuitOnLastWindowClosed);
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -15,6 +15,7 @@ All notable changes to this project are documented in this file.
 - Fixed ListView not updating when model changes (#3125)
 - Fixed TextInput on Plasma/Wayland receiveng many empty events causing selection to be cleared (#4148)
 - Added API to programmatically show a window in fullscreen mode (C++/Rust: `Window::set_fullscreen(bool)`, Node.js: `window.fullscreen`)
+- Added API to keep the event loop alive when the last window is closed (#1499). (Rust: `slint::run_event_loop_until_quit(); C++: argument to `slint::run_event_loop()`)
```
Comment (tronical) [inline CHANGELOG.md:18]:
```suggestion
- Added API to keep the event loop alive when the last window is closed (#1499). (Rust: `slint::run_event_loop_until_quit()`; C++: argument to `slint::run_event_loop()`)
```

**Review by tronical — APPROVED** [review]

**Pull request merged**